### PR TITLE
HDDS-6020: EC: Provide set replicationConfig option to bucket

### PR DESCRIPTION
**Pull request by umamaheswararao**
## What changes were proposed in this pull request?

Added the default replication setting flexibility via SetBucketPropertyRequest. I will raise separate issue to expose this option via CLI.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6020

## How was this patch tested?

Added tests.

### CODE REVIEW

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java [hunk 1]
```diff
@@ -704,6 +704,21 @@ public void setBucketQuota(String volumeName, String bucketName,
...
+  public void setReplicationConfig(
+      String volumeName, String bucketName, ReplicationConfig replicationConfig)
+      throws IOException {
+    verifyVolumeName(volumeName);
+    verifyBucketName(bucketName);
+    Preconditions.checkNotNull(replicationConfig);
+    OmBucketArgs.Builder builder = OmBucketArgs.newBuilder();
+    builder.setVolumeName(volumeName)
+        .setBucketName(bucketName)
+        .setDefaultReplicationCOnfig(
```
Comment (sodonnel) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:717]: There seems to be a typo in ".setDefaultReplicationCOnfig" where the "O" in config is in uppercase. Not sure if this is a pre-existing problem, as I don't see this method defined in this PR.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmBucketArgs.java [hunk 2]
```diff
@@ -199,15 +218,24 @@ public Builder setQuotaInNamespace(long quota) {
       return this;
     }
 
+    public Builder setDefaultReplicationCOnfig(
```
Comment (sodonnel) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmBucketArgs.java:221]: Ah, this is where the new method is defined I mentioned above. Typo in the method name COnfig -> Config.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java [hunk 3]
```diff
@@ -3753,4 +3757,65 @@ public void testOverWriteKeyWithAndWithOutVersioning() throws Exception {
...
   }
+
+  @Test
+  public void testSetECReplicationConfigOnBucket()
+      throws IOException {
+    String volumeName = UUID.randomUUID().toString();
+    store.createVolume(volumeName);
+    OzoneVolume volume = store.getVolume(volumeName);
+    OzoneBucket bucket = getBucket(volume);
+    Assert.assertEquals(bucket.getName(), bucket.getName());
```
Comment (sodonnel) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java:3768]: `Assert.assertEquals(bucket.getName(), bucket.getName());`

I don't think this assert is needed as its just checking the same variable equals itself.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java [hunk 4]
```diff
@@ -3753,4 +3757,65 @@ public void testOverWriteKeyWithAndWithOutVersioning() throws Exception {
...
+    Assert.assertEquals(
+        new ECReplicationConfig(3, 2, ECReplicationConfig.EcCodec.RS, 1024),
+        bucket.getReplicationConfig());
+
+    //Reset replication config back.
+    bucket.setReplicationConfig(currentReplicationConfig);
+  }
+
+  @Test
+  public void testSetRatisReplicationConfigOnBucket()
```
Comment (sodonnel) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java:3788]: Is it worth just merging this test into the one above? Eg set the config to EC, check it, then set it to Ratis and check it, and then restore to default and check it?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java [hunk 5]
```diff
@@ -3753,4 +3757,65 @@ public void testOverWriteKeyWithAndWithOutVersioning() throws Exception {
...
+      throws IOException {
+    String volumeName = UUID.randomUUID().toString();
+    store.createVolume(volumeName);
+    OzoneVolume volume = store.getVolume(volumeName);
+    OzoneBucket bucket = getBucket(volume);
+    Assert.assertEquals(bucket.getName(), bucket.getName());
```
Comment (sodonnel) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java:3794]: Same comment as above, I don't think this assert for `bucket.getName(), bucket.getName()` is needed.

**Comment by sodonnel** [comment]
This change looks mostly good to me. Just a couple of minor comments to address.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java [hunk 6]
```diff
@@ -3753,4 +3757,47 @@ public void testOverWriteKeyWithAndWithOutVersioning() throws Exception {
...
+    Assert.assertEquals(ecReplicationConfig, bucket.getReplicationConfig());
+
+    RatisReplicationConfig retisReplicationConfig =
+        new RatisReplicationConfig(HddsProtos.ReplicationFactor.THREE);
+    bucket.setReplicationConfig(retisReplicationConfig);
+
+    // Get the bucket and check the updated config.
+    bucket = volume.getBucket(bucket.getName());
+
+    Assert.assertEquals(retisReplicationConfig, bucket.getReplicationConfig());
```
Comment (sodonnel) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java:3788]: One small typo here retis -> ratis.

Comment (umamaheswararao) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneRpcClientAbstract.java]: Thanks for catching. I have just fixed it.

**Review by sodonnel — APPROVED** [review]
LGTM

**Pull request merged**